File: src/power_system.cc
#include "endfield_base/power_system.h"

#include <algorithm>
#include <cmath>
#include <unordered_map>

namespace endfield_base {
namespace {
    auto isCoveredByAnyPowerPole(
        const SimulationState& state,
        const FacilityInstance& target,
        const FacilityDefinition& targetDefinition
    ) -> bool {
        const auto targetCells = state.grid.getOccupiedCells(targetDefinition, target);

        for (const FacilityInstance& candidate : state.grid.getFacilities()) {
            const FacilityDefinition* candidateDefinition = state.catalog.findDefinition(candidate.definitionId);
            if (candidateDefinition == nullptr || !isPowerFacility(*candidateDefinition)) {
                continue;
            }

            const std::unique_ptr<PowerFacilityInterface> powerFacility =
                createPowerFacilityInterface(*candidateDefinition);
            if (powerFacility == nullptr) {
                continue;
            }

            const auto poleCells = state.grid.getOccupiedCells(*candidateDefinition, candidate);
            for (const GridPoint& targetCell : targetCells) {
                for (const GridPoint& poleCell : poleCells) {
                    const int manhattanDistance = std::abs(targetCell.x - poleCell.x) + std::abs(
                                                      targetCell.y - poleCell.y);
                    if (manhattanDistance <= powerFacility->getPowerRange(*candidateDefinition)) {
                        return true;
                    }
                }
            }
        }

        return false;
    }

    bool isGeneratingFacilityActive(const FacilityDefinition& definition, const FacilityInstance& instance) {
        if (definition.powerGeneration <= 0.0) {
            return false;
        }

        if (definition.category == FacilityCategory::Generator) {
            return true;
        }

        if (isThermalMachine(definition)) {
            return hasAnyStoredItems(instance);
        }

        return true;
    }
}  // namespace

PowerReport PowerSystem::evaluatePower(const SimulationState& state) {
    PowerReport report;

    struct PendingConsumer {
        std::size_t facilityStateIndex = 0;
        double requestedPower = 0.0;
    };

    std::vector<PendingConsumer> coveredConsumers;

    for (const FacilityInstance& instance : state.grid.getFacilities()) {
        const FacilityDefinition* definition = state.catalog.findDefinition(instance.definitionId);
        if (definition == nullptr) {
            continue;
        }

        FacilityPowerState facilityState;
        facilityState.instanceId = instance.instanceId;
        facilityState.requestedPower = definition->powerUsage;
        facilityState.covered = !requiresExternalPower(*definition)
                                || isCoveredByAnyPowerPole(state, instance, *definition);

        if (isGeneratingFacilityActive(*definition, instance)) {
            report.totalGeneration += definition->powerGeneration;
        }

        if (definition->powerUsage > 0.0) {
            report.totalRequested += definition->powerUsage;
        }

        if (!requiresExternalPower(*definition)) {
            facilityState.powered = true;
            facilityState.servedPower = definition->powerUsage;
            report.totalServed += definition->powerUsage;
        } else if (facilityState.covered) {
            coveredConsumers.push_back({report.facilityStates.size(), definition->powerUsage});
        }

        report.facilityStates.push_back(facilityState);
    }

    std::sort(
        coveredConsumers.begin(),
        coveredConsumers.end(),
        [&report] (const PendingConsumer& left, const PendingConsumer& right) {
            return report.facilityStates.at(left.facilityStateIndex).instanceId
                   < report.facilityStates.at(right.facilityStateIndex).instanceId;
        }
    );

    double remainingGeneration = report.totalGeneration;
    for (const PendingConsumer& consumer : coveredConsumers) {
        FacilityPowerState& facilityState = report.facilityStates.at(consumer.facilityStateIndex);
        if (remainingGeneration >= consumer.requestedPower) {
            facilityState.powered = true;
            facilityState.servedPower = consumer.requestedPower;
            remainingGeneration -= consumer.requestedPower;
            report.totalServed += consumer.requestedPower;
        }
    }

    return report;
}
}  // namespace endfield_base

File: src/EndfieldBase.Core/Power/PowerSystem.cs
using EndfieldBase.Core.Models;

namespace EndfieldBase.Core.Power;

public sealed class FacilityPowerState
{
    public int InstanceId { get; set; }
    public bool Covered { get; set; }
    public bool Powered { get; set; }
    public double RequestedPower { get; set; }
    public double ServedPower { get; set; }
}

public sealed class PowerReport
{
    public double TotalGeneration { get; set; }
    public double TotalRequested { get; set; }
    public double TotalServed { get; set; }
    public List<FacilityPowerState> FacilityStates { get; } = [];
}

public static class PowerSystem
{
    public static PowerReport EvaluatePower(SimulationState state) {
        PowerReport report = new();
        List<(int FacilityStateIndex, double RequestedPower)> coveredConsumers = new();

        foreach (var instance in state.Grid.GetFacilities()) {
            var definition = state.Catalog.FindDefinition(instance.DefinitionId);
            if (definition is null) continue;

            FacilityPowerState facilityState = new()
                {
                    InstanceId = instance.InstanceId,
                    RequestedPower = definition.PowerUsage,
                    Covered = !FacilityRules.RequiresExternalPower(definition) ||
                              IsCoveredByAnyPowerPole(state, instance, definition)
                };

            if (IsGeneratingFacilityActive(definition, instance)) report.TotalGeneration += definition.PowerGeneration;

            if (definition.PowerUsage > 0.0) report.TotalRequested += definition.PowerUsage;

            if (!FacilityRules.RequiresExternalPower(definition)) {
                facilityState.Powered = true;
                facilityState.ServedPower = definition.PowerUsage;
                report.TotalServed += definition.PowerUsage;
            }
            else if (facilityState.Covered) {
                coveredConsumers.Add((report.FacilityStates.Count, definition.PowerUsage));
            }

            report.FacilityStates.Add(facilityState);
        }

        coveredConsumers.Sort((left, right) =>
            report.FacilityStates[left.FacilityStateIndex].InstanceId
                .CompareTo(report.FacilityStates[right.FacilityStateIndex].InstanceId));

        var remainingGeneration = report.TotalGeneration;
        foreach (var consumer in coveredConsumers) {
            var facilityState = report.FacilityStates[consumer.FacilityStateIndex];
            if (remainingGeneration < consumer.RequestedPower) continue;

            facilityState.Powered = true;
            facilityState.ServedPower = consumer.RequestedPower;
            remainingGeneration -= consumer.RequestedPower;
            report.TotalServed += consumer.RequestedPower;
        }

        return report;
    }

    private static bool IsCoveredByAnyPowerPole(
        SimulationState state,
        FacilityInstance target,
        FacilityDefinition targetDefinition) {
        List<GridPoint> targetCells = state.Grid.GetOccupiedCells(targetDefinition, target);
        foreach (var candidate in state.Grid.GetFacilities()) {
            var candidateDefinition = state.Catalog.FindDefinition(candidate.DefinitionId);
            if (candidateDefinition is null || !FacilityRules.IsPowerFacility(candidateDefinition) ||
                candidateDefinition.PowerRange <= 0) continue;

            List<GridPoint> poleCells = state.Grid.GetOccupiedCells(candidateDefinition, candidate);
            foreach (var targetCell in targetCells)
                foreach (var poleCell in poleCells) {
                    var manhattanDistance = Math.Abs(targetCell.X - poleCell.X) + Math.Abs(targetCell.Y - poleCell.Y);
                    if (manhattanDistance <= candidateDefinition.PowerRange) return true;
                }
        }

        return false;
    }

    private static bool IsGeneratingFacilityActive(FacilityDefinition definition, FacilityInstance instance) {
        if (definition.PowerGeneration <= 0.0) return false;

        if (definition.Category == FacilityCategory.Generator) return true;

        return !FacilityRules.IsThermalMachine(definition) || FacilityRules.HasAnyStoredItems(instance);
    }
}